icu4c/source/test/intltest/rbbitst.cpp

```diff
@@ -4098,53 +4287,72 @@ void RBBITest::RunMonkey(BreakIterator *bi, RBBIMonkeyKind &mk, const char *name
-                        errorText.append("\\");
+                        format =  "--> %3i : ";  // Error position
                     }
-                    if (c < 0x10000) {
-                        errorText.append("\\u");
-                        for (bn=12; bn>=0; bn-=4) {
-                            errorText.append(hexChars.charAt((c>>bn)&0xf));
-                        }
-                    } else {
-                        errorText.append("\\U");
-                        for (bn=28; bn>=0; bn-=4) {
-                            errorText.append(hexChars.charAt((c>>bn)&0xf));
-                        }
+                    sprintf(buffer, format.c_str(), ci);
+                    errorText.append(buffer);
+
+                    // BMP or SMP character in hex
+                    format = "    \\u%04x";
+                    if (c >= 0x10000) {
+                        format = "\\U%08x";
                     }
-                    ci = testText.moveIndex32(ci, 1);
-                }
-                errorText.append("\\");
-                errorText.append("</data>\n");
+                    sprintf(buffer, format.c_str(), c);
+                    errorText.append(buffer);
+
+                    // Indicate reference and actual break status
+                    errorText.append(expectedBreaks[ci] == 0 ? "  . " : "  | ");
+                    errorText.append(currentBreakData[ci] == 0 ? " . " : " | ");
+                    
+                    // Get the class name and character name for the character.
+                    const unsigned long classIndex = mk.classIndexFromTestTextClassIndex(c);
+                    char cName[200];
+                    UErrorCode status = U_ZERO_ERROR;
+                    u_charName(c, U_EXTENDED_CHAR_NAME, cName, sizeof(cName), &status);
 
-                // Output the error
-                char  charErrorTxt[500];
-                UErrorCode status = U_ZERO_ERROR;
-                errorText.extract(charErrorTxt, sizeof(charErrorTxt), NULL, status);
-                charErrorTxt[sizeof(charErrorTxt)-1] = 0;
-                const char *badLocale = bi->getLocaleID(ULOC_ACTUAL_LOCALE, status);
+                    // The applied rule string
+                    const std::string appliedRule = mk.getAppliedRule(ci);
 
-                errln("%s break monkey test error [%s].  %s. Operation = %s; Random seed = %d;  buf Idx = %d\n%s",
-                    name, badLocale, (expectedBreaks[i]? "break expected but not found" : "break found but not expected"),
-                    errorType, seed, i, charErrorTxt);
+                    sprintf(buffer, "%15s  %20s -->>>  %20s", 
+                        charClassNames[classIndex].c_str(), cName, appliedRule.c_str());
+                    errorText.append(buffer);
+
+                    // Move to next character position
+                    int ci_old = ci;
+                    ci = testText.moveIndex32(ci, 1);
+                    if (ci == ci_old) {
```

Comment:
It seems odd that this check is necessary, like something is incorrect in the computation of endContext. If everything is working correctly, it should be possible to move "testText.moveIndex32(ci, 1)" into the for () statement itself. Or, if that's not practical for some reason, it should happen after outputting the error.

Reply:
OK, the problem is that moveIndex32() results are pinned to the string length, and endContext may be set to the string length, and the loop exit is ci > endContext.

Maybe recast the loop,
    
    for (ci=startContext; ; ci = testText.moveIndex32(ci, 1)) {
        ...
        // Output the error
        ...
        if (ci >= endContext) { break; }
    }

Reply:
Done.